### PR TITLE
feat: authentication flow

### PR DESCRIPTION
**Pull request by kayra1**
# Description

This PR introduces authentication to the frontend of NMS. The user will either be redirected to the initialize endpoint or login endpoint if they haven't logged in. Admin users have access to the users panel, which allows them to create, delete or update the password of existing user accounts.

For testing, please use the webconsole version that includes the authentication changes.

# Checklist:

- [x] My code follows the [style guidelines](/CONTRIBUTING.md) of this project
- [x] I have performed a self-review of my own code.
- [x] I have made corresponding changes to the documentation.
- [x] I have added tests that validate the behaviour of the software.
- [x] I validated that new and existing unit tests pass locally with my changes.
- [x] Any dependent changes have been merged and published in downstream modules.

### CODE REVIEW

**Review by gruyaume — CHANGES_REQUESTED** [review]
As soon as I initialize the admin user (or try to log in) I get redirected to the login page. This is by cloning your fork and going through the CONTRIBUTING.md guidelines. 

I believe I'm properly logged in but something redirects me to the login page regardless.

image

File: utils/auth.tsx [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+    useEffect(() => {
+        if (token) {
+            let userObject = jwtDecode(token) as User
+            userObject.authToken = token
+            setUser(userObject);
+        } else {
+            setUser(null)
+            router.push('/login');
+        }
+    }, [token]);
```
Comment (gruyaume) [inline utils/auth.tsx:31]: React Hook useEffect has a missing dependency: 'router'. Either include it or remove the dependency array

**Comment by gruyaume** [comment]
> As soon as I initialize the admin user (or try to log in) I get redirected to the login page. This is by cloning your fork and going through the CONTRIBUTING.md guidelines.
> 
> I believe I'm properly logged in but something redirects me to the login page regardless.
> 
> image

Actually, it seems like the token does not get added to my cookies.

**Comment by gruyaume** [comment]
> > As soon as I initialize the admin user (or try to log in) I get redirected to the login page. This is by cloning your fork and going through the CONTRIBUTING.md guidelines.
> > I believe I'm properly logged in but something redirects me to the login page regardless.
> > image
> 
> Actually, it seems like the token does not get added to my cookies.

This is caused by the `secure` option when setting the cookie. It is set to True but the CONTRIBUTING.md and make file lead to an environment that does not use https.

File: app/(nms)/users/page.tsx
Comment (gruyaume) [inline app/(nms)/users/page.tsx:1]: The action button is misaligned:

image

File: app/(nms)/users/modals.tsx
Comment (gruyaume) [inline app/(nms)/users/modals.tsx:1]: When I create a used that already exists, I am not informed that the operation didn't succeed, we should likely add a error notification here.

image

File: app/(nms)/users/page.tsx [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+            />
+          ),
+        },
+      ],
+    };
+  });
+
+  return (
+    <>
+      <PageHeader title={`NMS Accounts (${users.length})`}>
```
Comment (gruyaume) [inline app/(nms)/users/page.tsx:71]:
```suggestion
      <PageHeader title={`Users (${users.length})`}>
```

We are interchangibly using "accounts" and "users", let's make sure to always use the same word when we identify this concept.

**Review by gruyaume — CHANGES_REQUESTED** [review]

**Comment by kayra1** [comment]
> This is caused by the `secure` option when setting the cookie. It is set to True but the CONTRIBUTING.md and make file lead to an environment that does not use https.

The code to add the token to the cookies is identical to the one in Notary, and I can't recreate this issue locally even in a fresh VM. @patriciareinoso had the same issue. If it's possible I would appreciate hopping on a quick call to debug.

**Comment by kayra1** [comment]
> This is caused by the `secure` option when setting the cookie. It is set to True but the CONTRIBUTING.md and make file lead to an environment that does not use https.

i usually have an ssh tunnel into the environment, and localhost is excepted from the `Secure` cookie. I've removed the secure option, we can reintroduce it when making the change to move to https.

File: app/(nms)/users/page.tsx (hunk 2 above)
Comment (kayra1) [reply app/(nms)/users/page.tsx]: Updated

File: app/(nms)/users/page.tsx
Comment (kayra1) [reply app/(nms)/users/page.tsx:1]: this took me 3 hours

File: utils/auth.tsx [hunk 3]
```diff
@@ -0,0 +1,44 @@
+"use client"
+
+import { createContext, useContext, useState, useEffect, Dispatch, SetStateAction } from 'react';
```
[the remark this answers is not in the document]
Comment (kayra1) [reply utils/auth.tsx]: gone

File: app/(nms)/users/modals.tsx
Comment (kayra1) [reply app/(nms)/users/modals.tsx:1]: done

**Review by gruyaume — APPROVED** [review]

**Review by patriciareinoso — CHANGES_REQUESTED** [review]
After creating a network slice or a subscriber, I need to refresh the page to see it on the list. But this only happens with the first one. I think this was raised as an issue that was closed some time agou because it disappeared but now seems to be back. @gruyaume

**Review by patriciareinoso — CHANGES_REQUESTED** [review]
If I log in as a not admin user and I try to reach <ip>:5000/users on the browser address even if the `Users` button does not exist. It says loading and keep trying to the the users endpoint (http://10.97.137.188:5000/config/v1/account). Could it be possible to block the user? show something like `page does not exist` ? 

image

**Comment by kayra1** [comment]
> If I log in as a not admin user and I try to reach :5000/users on the browser address even if the `Users` button does not exist. It says loading and keeps trying to reach the users endpoint (http://10.97.137.188:5000/config/v1/account). Could it be possible to block the user? show something like `page does not exist` ?
> 
> image

how about redirecting to "/" for now? I want to pay special attention to having a 404 page, and preferably would like to do that after converting everything to canonical react components.

**Comment by patriciareinoso** [comment]
> > If I log in as a not admin user and I try to reach :5000/users on the browser address even if the `Users` button does not exist. It says loading and keeps trying to reach the users endpoint (http://10.97.137.188:5000/config/v1/account). Could it be possible to block the user? show something like `page does not exist` ?
> > image
> 
> how about redirecting to "/" for now? I want to pay special attention to having a 404 page, and preferably would like to do that after converting everything to canonical react components.

sure, that works

**Comment by patriciareinoso** [comment]
> After creating a network slice or a subscriber, I need to refresh the page to see it on the list. But this only happens with the first one. I think this was raised as an issue that was closed some time agou because it disappeared but now seems to be back. @gruyaume

as discussed, this will be tracked in a task (TELCO-1477) since it's not related to authentication.

**Review by patriciareinoso — APPROVED** [review]

**Pull request merged**